Add linter for a single element for loop

patrickelectric opened the pull request:
changelog: Fixes #1540, check for vectors that contain a single element in a for loop

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @flip1995 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

patrickelectric commented on clippy_lints/src/loops.rs:
```diff
@@ -1684,6 +1711,27 @@ fn check_for_loop_over_map_kv<'tcx>(
+                    cx,
+                    SINGLE_ELEMENT_LOOP,
+                    expr.span,
+                    "Do not iterate over a single element array.",
```
thanks!

patrickelectric commented:
ping @flip1995

patrickelectric commented:
ping @flip1995

flip1995 commented:
Sorry for taking so long. I currently have 15 PRs I'm actively reviewing/triaging besides doing other organizational stuff for Clippy. I usually review on weekends, so pinging doesn't really help to speed up my reviews.

bors commented:
:umbrella: The latest upstream changes (presumably #5727) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

flip1995 commented on tests/ui/single_element_loop.rs:
```diff
@@ -0,0 +1,9 @@
+// Tests from for_loop.rs that don't have suggestions
+
+#[warn(clippy::single_element_loop)]
```
We should definitely add a suggestion here. Let me know if you need help with that. You can take a look at `span_lint_and_sugg` usages to see how to do this.

patrickelectric replied:
Thanks for the suggestion, now it's using `span_lint_and_sugg`

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -1684,6 +1711,27 @@ fn check_for_loop_over_map_kv<'tcx>(
 }
 
+fn check_for_single_element_loop<'tcx>(
+    cx: &LateContext<'tcx>,
+    _: &'tcx Pat<'_>,
+    arg: &'tcx Expr<'_>,
+    _: &'tcx Expr<'_>,
+    _: &'tcx Expr<'_>,
```
No need to add all arguments, if they are unused.

patrickelectric replied:
Oh, I was just following the style of the other functions, removed!

flip1995 commented on clippy_lints/src/loops.rs:
````diff
@@ -453,6 +453,31 @@ declare_clippy_lint! {
+    /// println!("{}", item);
+    /// ```
+    pub SINGLE_ELEMENT_LOOP,
+    style,
````
```suggestion
    complexity,
```

patrickelectric replied:
Done!

flip1995 reviewed (commented):
Thanks for the contribution and welcome to Clippy!

patrickelectric commented:
Hi @flip1995, sorry for the ping, is just something that I do after a couple of days if a PR does not get a review, but since you mainly review PRs during the weekends I'll stop ping here. Sorry for the bothering! 
I just fixed the requested changes and rebased over the current master branch, let me know if anything else is necessary.
Have a great weekend!

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
+        if let ExprKind::AddrOf(BorrowKind::Ref, _, ref expr) = arg.kind;
+        if let PatKind::Binding(.., target, _) = pat.kind;
+        if let ExprKind::Array(ref expr_list) = expr.kind;
+        if expr_list.len() == 1;
+        if let ExprKind::Path(ref list_item) = expr_list[0].kind;
```
```suggestion
        if let [expr] = expr_list; // or a more fitting name than "expr"
        if let ExprKind::Path(ref list_item) = expr.kind;
```

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
+        if let ExprKind::Array(ref expr_list) = expr.kind;
+        if expr_list.len() == 1;
+        if let ExprKind::Path(ref list_item) = expr_list[0].kind;
+        if let Some(list_tem_name) = single_segment_path(list_item).map(|ps| ps.ident.name);
```
```suggestion
        if let Some(list_item_name) = single_segment_path(list_item).map(|ps| ps.ident.name);
```

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
+                cx,
+                SINGLE_ELEMENT_LOOP,
+                expr.span,
+                "do not iterate over a single element array",
+                "try",
```
```suggestion
                "`for` loop over a single element",
                "try",
```

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -1865,6 +1893,29 @@ fn check_for_loop_over_map_kv<'tcx>(
+            span_lint_and_sugg(
+                cx,
+                SINGLE_ELEMENT_LOOP,
+                expr.span,
```
You should use the span from the start of the for loop to the end of `expr`. You can do this, by also passing the `expr` from the `check_for_loop` function and then use the span `for_expr.span.with_hi(expr.span.hi())`.

flip1995 commented on tests/ui/single_element_loop.rs:
```diff
@@ -0,0 +1,9 @@
+// Tests from for_loop.rs that don't have suggestions
```
```suggestion
// run-rustfix
// Tests from for_loop.rs that don't have suggestions
```

patrickelectric commented:
Needs https://github.com/rust-lang/rust-clippy/pull/6191 to compile.

patrickelectric commented:
ping @flip1995

flip1995 commented:
@bors r+

Thanks!

Sorry for taking so long. I'll start a new job in a new town next month and have little time for reviews currently. I hope it settles down soon and I have more time and energy for Clippy again.

bors commented:
:pushpin: Commit ba1ca19c3bec20401a4cb13e5186c4c5952e94cc has been approved by `flip1995`

<!-- @bors r=flip1995 ba1ca19c3bec20401a4cb13e5186c4c5952e94cc -->
<!-- homu: {"type":"Approved","sha":"ba1ca19c3bec20401a4cb13e5186c4c5952e94cc","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit ba1ca19c3bec20401a4cb13e5186c4c5952e94cc with merge e0e617adaa89c759beccf74bc0f9eafaed44de22...
<!-- homu: {"type":"BuildStarted","head_sha":"ba1ca19c3bec20401a4cb13e5186c4c5952e94cc","merge_sha":"e0e617adaa89c759beccf74bc0f9eafaed44de22"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1304466561), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1304465343), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1304503574)
Approved by: flip1995
Pushing e0e617adaa89c759beccf74bc0f9eafaed44de22 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1304503574","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1304466561","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1304465343"},"merge_sha":"e0e617adaa89c759beccf74bc0f9eafaed44de22"} -->

patrickelectric commented:
Hi @flip1995, thank you for the review, I was just a bit worried about the hacktoberfest schedule.
Congratulations for the new job and I hope all is well in the coming weeks!

flip1995 commented:
IIUC for hacktoberfest only the opening date of the PR is relevant. :wink:
